src/i_video.c

```diff
@@ -986,6 +986,14 @@ void I_SetGammaTable (void)
 void I_SetPalette (byte *doompalette)
 {
     int i;
+    static boolean gamma2table_set = false;
```

Comment:
Sorry for being pedantic, but somehow I don't like the fact that `I_SetPalette()` has to check each time if the gamma values have already been calculated. They should be calculated once as early as possible and then be taken as granted. They replace static values after all.

Reply:
This was primary reason of confusion. My initial idea was to put `I_SetPalette()` to `I_InitGraphics()` below [this line](https://github.com/fabiangreffrath/crispy-doom/blob/gamma/src/i_video.c#L1751) and then do not make any farther calls to `I_SetGammaTable()` at all. And it's working fine for palleted render, for true color, however, I'm starting to get black screen. 

I don't understand why. Such call to `I_SetGammaTable()` is not limited to `#ifndef CRISPY_TRUECOLOR`, and gamma values are must be set. What else missing then?

Reply:
@fabiangreffrath, allow me please to clarify about last commit https://github.com/fabiangreffrath/crispy-doom/pull/1121/commits/c51afc83fb24064c580bad1019108db7dcbae79b:
- For palleted render it seems to be correct: we defining gamma tables right before applying palette. It's "calculate first, draw second" logics.
- For true color... It's just disgusting. It does not feels to be the right place for calling `I_SetGammaTable()`, but this is the only way I was able to make it work w/o extra checks. Even if `I_SetGammaTable()` is moved outside of `#ifndef CRISPY_TRUECOLOR` macroses in `I_InitGraphics()`, it simply doesn't work, matters no where exactly it is placed. I'm completelly out of ideas why - `gamma2table` does not seems to be affected by `CRISPY_TRUECOLOR` in any ways and it's not palette related thing. So what could it be then? Kaboom! 🤯

Reply:
Alright, your fix is all fine. Let me try to explain why. :wink: 

In the non-truecolor case, `I_SetGammaTable()` is called from `I_InitGraphics()` right before `PLAYPAL` is loaded and applied to `screenbuffer->format->palette`. This is the first time the gamma table is actually used in this case.

As you can see in `src/doom/d_main.c`, `I_InitGraphics()` is called rather late. Very late, actually, in the never-returning `D_DoomLoop()`, right before entering the `D_RunFrame()` loop. That is, *everything else* in this game is already initialized before the graphics are set up, which in turn is the earliest situation in which the gamma tables is used. But only in the non-truecolor case!

Let's stay in `src/doom/d_main.c` and have a look at the single large `D_DoomMain()` function. You'll see that it calls a lot of initialization stuff, among these `R_Init`, *before* it calls `D_DoomLoop()` in the last line of that function. This means that `R_Init()` and thus `R_InitColormaps()` is always called before `I_InitGraphics()` and so the gamma table is never set up at that point.

Now, why does it make a difference for truecolor and non-truecolor rendering? Because in the case of the palette renderer, we just *load the `colormaps` from the `COLORMAP` lump* and be done with it. However, for the truecolor renderer we need to *calculate* the `colormaps` entries and have to take the current gamma level into account. That's why this function will calculate an all-black `colormap` if the gamma table hasn't been calculated before. And so, it's perfectly right to call `I_SetGammaTable()` right before `R_InitColormaps()` in the truecolor case and much later during `I_InitGraphics()` in the non-truecolor case.

Reply:
Thank you very much for this not explanation, but whole guide, a road map. I was walking this path strictly forward, while small close look backward was needed. True color's `if (!gamma2table)` was a small beacon which led to the right path, and this whole complicated structure of calls was the real reason behind confusion.

So yeah, `I_InitGraphics()` is called after `R_Init()`, it is clear now. This is also making next perfectly clear: we need to get rid `#ifndef CRISPY_TRUECOLOR` macros and make both renders happy. How? Simple! We need exactly one call for `I_SetGammaTable()` right before `R_InitColormaps()` but without any marcoses. True color render wants it right there, and palleted render will be ready for it when it will be calling `I_InitGraphics()`.

The only personal nitpick is having `I_` upper level function in the in `R_` lower level function, but it have to be `I_` as it placed in upper level and it's shared for all four games. Not really nice, but makes sense for understanding where `I_SetGammaTable()` is located.

We can simplify even farther by having exactly one call for `I_SetGammaTable()`, but then it should be placed in `int main()` function which is **a-b-s-o-l-u-t-e-l-y no go**, it simply doesn't belong there matters no how simple it is!